File: services/dataService.ts
import { Card, TradePlan, BundledCard } from "../types";

// Use relative URL in production, localhost in development
const API_URL = import.meta.env.DEV ? 'http://localhost:3001/api' : '/api';

// Helper to get auth headers
const getAuthHeaders = async (getIdToken: () => Promise<string | null>) => {
  const token = await getIdToken();
  return {
    'Content-Type': 'application/json',
    ...(token ? { 'Authorization': `Bearer ${token}` } : {})
  };
};

export const dataService = {
  async getCards(getIdToken: () => Promise<string | null>): Promise<Card[]> {
    try {
      const headers = await getAuthHeaders(getIdToken);
      const res = await fetch(`${API_URL}/cards`, { headers });
      if (!res.ok) throw new Error('Failed to fetch cards');
      return await res.json();
    } catch (e) {
      console.error("API Error", e);
      return [];
    }
  },

  async saveCard(card: Card, getIdToken: () => Promise<string | null>): Promise<Card> {
    const headers = await getAuthHeaders(getIdToken);
    const res = await fetch(`${API_URL}/cards`, {
      method: 'POST',
      headers,
      body: JSON.stringify(card)
    });
    return await res.json();
  },

  async deleteCard(id: string, getIdToken: () => Promise<string | null>): Promise<void> {
    const headers = await getAuthHeaders(getIdToken);
    await fetch(`${API_URL}/cards/${id}`, {
      method: 'DELETE',
      headers
    });
  },

  async updatePrice(id: string, newPrice: number, getIdToken: () => Promise<string | null>, dateStr?: string, platform?: string, parallel?: string, grade?: string, serialNumber?: string): Promise<Card | null> {
    const headers = await getAuthHeaders(getIdToken);
    console.log('[dataService] Sending updatePrice request:', { id, price: newPrice, date: dateStr, platform, parallel, grade, serialNumber });
    const res = await fetch(`${API_URL}/cards/${id}/price`, {
      method: 'POST',
      headers,
      body: JSON.stringify({ price: newPrice, date: dateStr, platform, parallel, grade, serialNumber })
    });
    if (!res.ok) return null;
    const result = await res.json();
    console.log('[dataService] updatePrice response:', result);
    return result;
  },

  async deletePriceEntry(id: string, priceDate: string, getIdToken: () => Promise<string | null>): Promise<Card | null> {
    const headers = await getAuthHeaders(getIdToken);
    console.log('[dataService] Sending deletePriceEntry request:', { id, priceDate });
    const res = await fetch(`${API_URL}/cards/${id}/price/${encodeURIComponent(priceDate)}`, {
      method: 'DELETE',
      headers
    });
    if (!res.ok) return null;
    const result = await res.json();
    console.log('[dataService] deletePriceEntry response:', result);
    return result;
  },

  async editPriceEntry(id: string, oldDate: string, newPrice: number, getIdToken: () => Promise<string | null>, newDate?: string, platform?: string, parallel?: string, grade?: string, serialNumber?: string): Promise<Card | null> {
    const headers = await getAuthHeaders(getIdToken);
    console.log('[dataService] Sending editPriceEntry request:', { id, oldDate, newPrice, newDate, platform, parallel, grade, serialNumber });
    const res = await fetch(`${API_URL}/cards/${id}/price/${encodeURIComponent(oldDate)}`, {
      method: 'PUT',
      headers,
      body: JSON.stringify({ price: newPrice, date: newDate, platform, parallel, grade, serialNumber })
    });
    if (!res.ok) return null;
    const result = await res.json();
    console.log('[dataService] editPriceEntry response:', result);
    return result;
  },

  // ===== Trade Plans API =====

  async getTradePlans(getIdToken: () => Promise<string | null>, status?: 'pending' | 'completed' | 'cancelled'): Promise<TradePlan[]> {
    try {
      const headers = await getAuthHeaders(getIdToken);
      const url = status ? `${API_URL}/trade-plans?status=${status}` : `${API_URL}/trade-plans`;
      const res = await fetch(url, { headers });
      if (!res.ok) throw new Error('Failed to fetch trade plans');
      return await res.json();
    } catch (e) {
      console.error("API Error fetching trade plans", e);
      return [];
    }
  },

  async getTradePlan(id: string, getIdToken: () => Promise<string | null>): Promise<TradePlan | null> {
    try {
      const headers = await getAuthHeaders(getIdToken);
      const res = await fetch(`${API_URL}/trade-plans/${id}`, { headers });
      if (!res.ok) return null;
      return await res.json();
    } catch (e) {
      console.error("API Error fetching trade plan", e);
      return null;
    }
  },

  async createTradePlan(plan: {
    planName: string;
    targetValue?: number;
    bundleCards: BundledCard[];
    cashAmount?: number;
    cashCurrency?: 'USD' | 'CNY';
    totalBundleValue: number;
    notes?: string;
  }, getIdToken: () => Promise<string | null>): Promise<TradePlan> {
    const headers = await getAuthHeaders(getIdToken);
    const res = await fetch(`${API_URL}/trade-plans`, {
      method: 'POST',
      headers,
      body: JSON.stringify(plan)
    });
    if (!res.ok) throw new Error('Failed to create trade plan');
    return await res.json();
  },

  async updateTradePlan(id: string, updates: Partial<TradePlan>, getIdToken: () => Promise<string | null>): Promise<TradePlan> {
    const headers = await getAuthHeaders(getIdToken);
    const res = await fetch(`${API_URL}/trade-plans/${id}`, {
      method: 'PUT',
      headers,
      body: JSON.stringify(updates)
    });
    if (!res.ok) throw new Error('Failed to update trade plan');
    return await res.json();
  },

  async deleteTradePlan(id: string, getIdToken: () => Promise<string | null>): Promise<void> {
    const headers = await getAuthHeaders(getIdToken);
    await fetch(`${API_URL}/trade-plans/${id}`, {
      method: 'DELETE',
      headers
    });
  },

  async completeTradePlan(id: string, transactionId: string, getIdToken: () => Promise<string | null>): Promise<TradePlan> {
    const headers = await getAuthHeaders(getIdToken);
    const res = await fetch(`${API_URL}/trade-plans/${id}/complete`, {
      method: 'POST',
      headers,
      body: JSON.stringify({ transactionId })
    });
    if (!res.ok) throw new Error('Failed to complete trade plan');
    return await res.json();
  },

  async migrateTradePlansCurrency(getIdToken: () => Promise<string | null>): Promise<{ message: string; updated: number }> {
    const headers = await getAuthHeaders(getIdToken);
    const res = await fetch(`${API_URL}/trade-plans/migrate-currency`, {
      method: 'POST',
      headers
    });
    if (!res.ok) throw new Error('Failed to migrate trade plans');
    return await res.json();
  }
};

File: mobile/lib/services/data_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/card_model.dart';

class DataService {
  // Use 10.0.2.2 for Android emulator, localhost for iOS simulator
  // TODO: Make this configurable based on environment
  static const String _baseUrl = 'http://localhost:3001/api';

  Future<Map<String, String>> _getHeaders(Future<String?> Function() getIdToken) async {
    final token = await getIdToken();
    return {
      'Content-Type': 'application/json',
      if (token != null) 'Authorization': 'Bearer $token',
    };
  }

  Future<List<CardModel>> getCards(Future<String?> Function() getIdToken) async {
    try {
      final headers = await _getHeaders(getIdToken);
      final response = await http.get(Uri.parse('$_baseUrl/cards'), headers: headers);

      if (response.statusCode == 200) {
        final List<dynamic> jsonList = json.decode(response.body);
        return jsonList.map((json) => CardModel.fromJson(json)).toList();
      } else {
        throw Exception('Failed to load cards: ${response.statusCode}');
      }
    } catch (e) {
      print('Error fetching cards: $e');
      return [];
    }
  }

  Future<CardModel> saveCard(CardModel card, Future<String?> Function() getIdToken) async {
    final headers = await _getHeaders(getIdToken);
    final response = await http.post(
      Uri.parse('$_baseUrl/cards'),
      headers: headers,
      body: json.encode(card.toJson()),
    );

    if (response.statusCode == 200 || response.statusCode == 201) {
      return CardModel.fromJson(json.decode(response.body));
    } else {
      throw Exception('Failed to save card: ${response.statusCode}');
    }
  }

  Future<void> deleteCard(String id, Future<String?> Function() getIdToken) async {
    final headers = await _getHeaders(getIdToken);
    final response = await http.delete(
      Uri.parse('$_baseUrl/cards/$id'),
      headers: headers,
    );

    if (response.statusCode != 200 && response.statusCode != 204) {
      throw Exception('Failed to delete card: ${response.statusCode}');
    }
  }

  Future<CardModel?> updatePrice(
    String id,
    double newPrice,
    Future<String?> Function() getIdToken, {
    String? dateStr,
    String? platform,
    String? parallel,
    String? grade,
    String? serialNumber,
  }) async {
    final headers = await _getHeaders(getIdToken);
    final body = {
      'price': newPrice,
      'date': dateStr,
      'platform': platform,
      'parallel': parallel,
      'grade': grade,
      'serialNumber': serialNumber,
    };

    final response = await http.post(
      Uri.parse('$_baseUrl/cards/$id/price'),
      headers: headers,
      body: json.encode(body),
    );

    if (response.statusCode == 200) {
      return CardModel.fromJson(json.decode(response.body));
    } else {
      return null;
    }
  }

  Future<CardModel?> deletePriceEntry(
    String id,
    String priceDate,
    Future<String?> Function() getIdToken,
  ) async {
    final headers = await _getHeaders(getIdToken);
    final encodedDate = Uri.encodeComponent(priceDate);
    final response = await http.delete(
      Uri.parse('$_baseUrl/cards/$id/price/$encodedDate'),
      headers: headers,
    );

    if (response.statusCode == 200) {
      return CardModel.fromJson(json.decode(response.body));
    } else {
      return null;
    }
  }

  Future<CardModel?> editPriceEntry(
    String id,
    String oldDate,
    double newPrice,
    Future<String?> Function() getIdToken, {
    String? newDate,
    String? platform,
    String? parallel,
    String? grade,
    String? serialNumber,
  }) async {
    final headers = await _getHeaders(getIdToken);
    final encodedDate = Uri.encodeComponent(oldDate);
    final body = {
      'price': newPrice,
      'date': newDate,
      'platform': platform,
      'parallel': parallel,
      'grade': grade,
      'serialNumber': serialNumber,
    };

    final response = await http.put(
      Uri.parse('$_baseUrl/cards/$id/price/$encodedDate'),
      headers: headers,
      body: json.encode(body),
    );

    if (response.statusCode == 200) {
      return CardModel.fromJson(json.decode(response.body));
    } else {
      return null;
    }
  }
}
